core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -565,8 +573,14 @@ private[spark] class ExecutorAllocationManager(
     } else {
       // We don't want to change our target number of executors, because we already did that
       // when the task backlog decreased.
-      client.killExecutors(executorIdsToBeRemoved.toSeq, adjustTargetNumExecutors = false,
-        countFailures = false, force = false)
+      if (conf.get(WORKER_DECOMMISSION_ENABLED)) {
+        val executorIdsWithoutHostLoss = executorIdsToBeRemoved.toSeq.map(
+          id => (id, ExecutorDecommissionInfo("spark scale down", false)))
+        client.decommissionExecutors(executorIdsWithoutHostLoss, adjustTargetNumExecutors = false)
+      } else {
+        client.killExecutors(executorIdsToBeRemoved.toSeq, adjustTargetNumExecutors = false,
```

Comment:
Actually there is a key big difference between `killExecutors` and `decommissionExecutors`: The former does the actual killing and the latter does 'graceful death': Graceful death waits for all the migrations etc to happen and then the executor exits. I think this may lead to an overcommit: The old executor is still being decommissioned and migrating things, and a new one could be spun up. The two executors are alive at the same time (perhaps on the same node). Would this cause issues ?

This problem does not exist for `killExecutors` so much: Kill executor will kill the executor soon enough and thus the overlap window is smaller.

Reply:
We're talking in standalone mode? Multiple executors on the same worker is a deprecated feature so I don't think we have to worry about supporting new functionality with it.

Reply:
Okay. It is a little bit unsettling to mer personally, that we don't have a timeout on the time that an executor would take to "migrate and finally die", but I guess we can live with that shortcoming for a while. It just reduces the effective cluster capacity for an indeterminate amount of time, which is what makes me slightly queasy about it.

Reply:
So given this is in the Spark planned scale down path that _seems_ to be ok (the goal is to eventually reduce the capacity).